Hw3

Neonik-dev opened the pull request:
На текущий момент сделал только 1 задание, без допа с клавиатурой. И у меня еще не реализована обертка над SDK. Я сделал основную логику для бота, прописал все команды и их выполнение.
Проверь, пожалуйста, пока то, что есть, чтобы я внес правки и дальше работал с корректным кодом.

Для создания и вызова команд, я использовал паттерн Команда. У меня есть InputHandler, в который передаются все updates из тг. Так же я создал Мапу, в которой хранятся все названия команд и сам объект этой команды. В InputHandler берется текст пользователя и ищет в мапе тот класс, который должен выполнится для этой команды и вызывается метод execute. 
Все команды наследуются от интерфейса BaseCommand. Таким образом у всех есть метод execute в котором и выполняется основная логика.

Neonik-dev committed:
Completed 1 and 2 task. Create handlers and error handler

Neonik-dev committed:
Merge branch 'master' of https://github.com/Neonik228/Tinkoff_project into hw2

Neonik-dev committed:
Fix inaccuracies. Create GitHubClient

Neonik-dev committed:
Update VisitorGitHub

Neonik-dev committed:
Create StackOverflow client

Neonik-dev committed:
Create Scheduler

Neonik-dev committed:
Fixed the returned controller type

Neonik-dev committed:
Fixed mistakes, created SiteConfig, which parsed application.property

Neonik-dev committed:
Fixed mistakes, created SiteConfig, which parsed application.property

Neonik-dev committed:
Created the basic logic for the bot.

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/controllers/GeneralExceptionHandler.java:
```diff
@@ -0,0 +1,39 @@
+
+        return ResponseEntity.badRequest().body(new ApiErrorResponse(
+                "Invalid request parameters",
+                "400",
```
В константочку бы

Neonik-dev replied:
Fix

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/exceptions/NotUniqueLinkException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.tinkoff.edu.java.bot.logic.exceptions;
+
+public class NotUniqueLinkException extends Exception {
```
Папку exception вынести в главную bot

Neonik-dev replied:
угу, сделал

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/commands/Commands.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.HashMap;
+
+public class Commands {
+    private static final HashMap<String, BaseCommand> COMMANDS = new HashMap<>();
```
Валидная реализация, но мне кажется немного муторная, не хочешь сделать через enum как-нибудь? Если нет, то нет, но аргументируй

Neonik-dev replied:
Подумал над этим. Создал enum `InitBaseCommands `и `InitReplyCommands`. В итоге получилось сделать 2 реализации:
1) Оставить класс `Commands`, в котором происходит инициализация моих Map из `enum.values()`. Выглядит красиво и читаемо, но + еще один класс.
2) Удалить класс `Commands`, а создание Map переложил на сами Enum. Получается, каждый Enum сам создает свою собственную Map. Поэтому там так много `static`. Вроде и звучит неплохо, но из-за этого Enum`ы немного вздулись, в плане кода.

Сейчас действует 2 способ, но если выбрать первый, то все со static из Enum можно будет удалить. И получатся очень компактные Enum`ы.

Еще были проблемы, что static Map инициализировался быстрее, чем создавался сами значения Enum, из-за этого, когда я запрашивал Map мне возвращался просто null, а не сама мапа. Решил проблему создав вложенные класс `private static final class GuaranteeInit`

Подскажи, какой в итоге способ использовать.

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/commands/InvalidCommand.java:
```diff
@@ -0,0 +1,11 @@
+
+public class InvalidCommand {
+    public static SendMessage execute(Message message) {
+        String text = "Такой команды не существует.\nПопробуйте воспользоваться командой /help";
```
С енамом такие константочки можно будет туда вынести например

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/utils/InputHandler.java:
```diff
@@ -0,0 +1,41 @@
+                sendMessage = baseCommand.execute(message);
+            } else if (message.replyToMessage() != null) {
+                ReplyCommand replyCommand = REPLY_COMMANDS.get(message.replyToMessage().text());
+                if (replyCommand == null) {
```
Можно ли избавиться от тройного ифа если просто заранее брать и делать bot.execute?

Neonik-dev replied:
Попробовал (`bot.execute(baseCommand.execute(message));`), стало сложнее читаться и не особо понятно что возвращает baseCommand. Также теперь ведь надо отлавливать ошибку
\+ проверку на `replyCommand == null `нужно делать в любом случае, иначе бот примет null и не заметит.

В итоге сделал метод `checkReply `+ `if` заменил на `try/catch`. Думаю, стало лучше

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/utils/ManagerCollection.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.ArrayList;
+
+public class ManagerCollection {
+    private static final ArrayList<URI> LINKS = new ArrayList<>();
```
Может сет тогда, если тебе нужно contains? Будет работать за лог н

Neonik-dev replied:
эх, точно

Neonik-dev committed:
Completed task 1, without an additional task, and 2.
Сreated the basic logic for the bot, wrote an HTTP-client for the scrapper application

Neonik-dev commented:
Сейчас сделал первое, второе и четвертое задание, пока без допа.

Есть пару вопросов.
1) Где делать валидацию ссылки? В модуле bot, и гарантировать, что в scrapper придет валидная ссылка, или наоборот в scrapper, проверять link? Или и там, и там лучше проверять?
2) В примере, где создают обёртку, наследуются от `AutoCloseable`. Но я не понимаю, как использовать его в моем случае. Я запускаю `ConnectingBot `в классе `BotApplication`. И если я там прописываю(есть комменты там) `try() {...}` то он выполняет `start() `и тут же `close()`. Потому что команда start() не выполняется постоянно, поэтому момент и выходит из `try`. И я вот в недоумении как и где мне прописать открытие бота, чтобы он, закрывался сам. Ведь start это не цикличная команда, она выполнилась и пошла дальше.
3) Я долго пытался прокинуть `ScrapperClient `в классы с командами бота, что бы сразу делать запросы по апи(чуть опередил задание). Но я не понимаю как это можно реализовать. Итак, вопрос в том, как мне заинжектить бин `ScrapperClient `в `InputHandler`.
4) Во всех тестах у меня используется `ParseChain.chain()`. Насколько я правильно сделал, что вынес его как static поле? Что-то мне подсказывает, что это не совсем верное решение.
5) Можно ли как-то возвращать методы для каждого Теста. Например, в каждом тесте мне нужно заново, допустим поднимать базу данных, и отдавать ее тестам, как это сделать? Я знаю про аннотацию BeforeEach/BeforeAll, но они не возвращают методы. Или также, постоянно обновлять static поле?

Еще я хотел написать свой `SendSimpleMessage`, который наследовался от `SendMessage`. Но почему то bot.execute() не хочет работать с моим наследником и просто не выполняет отправку сообщения. Для меня это пока загадка.

Neonik-dev committed:
Did some tests on link-parser

Neonik-dev committed:
add more tests

Neonik-dev committed:
add more tests

Neonik-dev commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/TgUpdaterLinkBot.java:
```diff
@@ -0,0 +1,38 @@
+
+    public TgUpdaterLinkBot(ApplicationConfig config) {
+        this.config = config;
+    }
```
На мой взгляд этот класс создан совсем не по IoC. Но я не совсем понимаю как это исправить. Мне нужно, чтобы `TelegramBot bot = TelegramBot(config.token())` и как передать это в конструктор? Еще и проблема в том, что при инициализации он уже использует токен с конфига, который только-только передается в конструктор.
\+ Проблемы с InputHandler

Neonik-dev replied:
Вроде пофиксил

Neonik-dev committed:
update bean

Neonik-dev committed:
New Init Bean

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/configuration/ClientConfig.java:
```diff
@@ -0,0 +1,16 @@
+@RequiredArgsConstructor
+public class ClientConfig {
+    private final ScrapperConfiguration scrapperConfig;
+    @Bean("ScrapperClient")
```
Бины называются с маленькой буквы

Neonik-dev replied:
Запомнил, спасибо

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/clients/dto/AddLinkRequest.java:
```diff
@@ -0,0 +1,6 @@
+
+import java.net.URI;
+
+public record AddLinkRequest(URI url) {
```
Все дто лучше выносить в один пакет dto в папке bot, а внутри папки dto уже создавать подпапки

Neonik-dev replied:
Точно-точно

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/utils/InputHandler.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.HashMap;
+
+public class InputHandler {
```
Не думаю, что это прям утильный класс

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/commands/InitBaseCommands.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.HashMap;
+
+@Getter
+public enum InitBaseCommands {
```
Енамы в отдельную папку все

Neonik-dev replied:
Странно, до этого тоже задумался над этим, но в stackOverFlow писали, что их нужно хранить там же, где и обычные классы, которые связаны с этим Enum. Но окей, исправил.

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/commands/InitBaseCommands.java:
```diff
@@ -0,0 +1,35 @@
+        static final HashMap<String, BaseCommand> COMMANDS = new HashMap<>();
+    }
+
+    static {
```
Зачем нам этот статик блок, если мы можем прям в COMMANDS написать valules().stream().forEach и т.д.

Neonik-dev replied:
Потому что если в классе GuaranteeInit значения Enum еще не создались и возвращает просто null при вызове values()

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/commands/ListCommand.java:
```diff
@@ -0,0 +1,32 @@
+
+@Getter
+public class ListCommand implements BaseCommand {
+    private final String description = "выводит список отслеживающих ссылок";
```
Вообще в тот енам с коммандами можно и description завезти, а то зачем он тут впринципе то

Neonik-dev replied:
Тоже колебался над этим решением, но решил, что это относится именно к классу команды. Ведь мы можем изменить имя команды на любое друго, но вот описание команды и то, что она делает всегда взаимосвязаны и не могут быть отдельно.

Но окей, соглашусь, что это немного убавит код, поэтому исправлю

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/wrapper/TgBot.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.List;
+
+public interface TgBot extends UpdatesListener{
```
Перекинь этоть интерфейс в папку к боту, тяжеловато находить разные классы у тебя

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/TgUpdaterLinkBot.java:
```diff
@@ -0,0 +1,37 @@
+    private final InputHandler inputHandler;
+
+    @Override
+    public void start() {
```
Так если ты в итоге не реаилзуешь AutoClosable зачем тебе тут start и close?

Neonik-dev replied:
`close `я использую в `@PreDestroy `у бина. А `start `действительно особой роли не играет, предлагаешь вынести содержимое `start `в конструктор? или как? Вроде логично, если буду использовать `start `при инициализации bean.

На счет `AutoClosable `я задал вопрос в самом начале, возможно не заметил:

В примере, где создают обёртку, наследуются от `AutoCloseable`. Но я не понимаю, как использовать его в моем случае. Если я запускаю `ConnectingBot ` в `try() {...}` то он выполняет `start()` и тут же выходит из try и выполняет `close()`. Потому что команда `start()` не выполняется постоянно, поэтому моментально и выходит из try. И я вот в недоумении как и где мне прописать открытие бота, чтобы он, закрывался сам. Ведь start это не цикличная команда, она выполнилась и пошла дальше.

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/logic/utils/InputHandler.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.HashMap;
+
+public class InputHandler {
+    private static final HashMap<String, BaseCommand> COMMANDS = InitBaseCommands.getCommands();
```
Лучше юзать интерфейсы, то есть писать private static final Map<String, BaseCommand>

Gram3r commented:
> Сейчас сделал первое, второе и четвертое задание, пока без допа.
> 
> Есть пару вопросов.
> 
> 1. Где делать валидацию ссылки? В модуле bot, и гарантировать, что в scrapper придет валидная ссылка, или наоборот в scrapper, проверять link? Или и там, и там лучше проверять?
> 2. В примере, где создают обёртку, наследуются от `AutoCloseable`. Но я не понимаю, как использовать его в моем случае. Я запускаю `ConnectingBot `в классе `BotApplication`. И если я там прописываю(есть комменты там) `try() {...}` то он выполняет `start() `и тут же `close()`. Потому что команда start() не выполняется постоянно, поэтому момент и выходит из `try`. И я вот в недоумении как и где мне прописать открытие бота, чтобы он, закрывался сам. Ведь start это не цикличная команда, она выполнилась и пошла дальше.
> 3. Я долго пытался прокинуть `ScrapperClient `в классы с командами бота, что бы сразу делать запросы по апи(чуть опередил задание). Но я не понимаю как это можно реализовать. Итак, вопрос в том, как мне заинжектить бин `ScrapperClient `в `InputHandler`.
> 4. Во всех тестах у меня используется `ParseChain.chain()`. Насколько я правильно сделал, что вынес его как static поле? Что-то мне подсказывает, что это не совсем верное решение.
> 5. Можно ли как-то возвращать методы для каждого Теста. Например, в каждом тесте мне нужно заново, допустим поднимать базу данных, и отдавать ее тестам, как это сделать? Я знаю про аннотацию BeforeEach/BeforeAll, но они не возвращают методы. Или также, постоянно обновлять static поле?
> 
> Еще я хотел написать свой `SendSimpleMessage`, который наследовался от `SendMessage`. Но почему то bot.execute() не хочет работать с моим наследником и просто не выполняет отправку сообщения. Для меня это пока загадка.

Я вот вообще не понял зачем тебе возвращать метод если честно

Neonik-dev committed:
Fix all mistakes

Neonik-dev committed:
Add wrapper SendSimpleMessage

Neonik-dev committed:
Updates link-parsers tests

Neonik-dev committed:
created tests on the InputHandler for an unknown command

Neonik-dev commented:
Все сделал (доп тоже), все пофиксил.
Следующие коммиты буду делать просто так, для полноты приложения, их в задании нет.

Neonik-dev committed:
created tests on the *Command

Neonik-dev committed:
created tests on the *Command

Neonik228 commented:
На текущий момент выполнил первые 2 задачи. На ум пришло 2 решения по поводу архитектуры базы данных (вот итоговая модель  https://ibb.co/DGvQ8jK). 
1) Сделать наследования таблиц от основной. В основной бы хранилась главная информация по типу имени, даты обновления. А в доп таблицы создать для каждого домена и там указать поля, которые хотим отслеживать (коммиты, комментарии и тд)  ведь они у всех разные.
2) Использовать 1 таблицу, а всю отслеживаемую информацию хранить в jsonb, и не создавать бесконечно много табличек (ведь у нас может быть очень много разных ссылок). Таким образом нужен будет всего 1 запрос для выборки нужных  данных.
Я выбрал 2 вариант. Ведь для каждого домена нужно создать таблицу, и всякий раз делать 2 запроса. Еще мне показалось трудность в том, что имена таблиц мы будем хардкодить, а не получать из запроса, чтобы сделать второй запрос. То есть если изменится имя таблицы, то придется идти и искать в коде, где оно использовалось и там тоже менять все. Во втором случае такого не будет. 
Буду рад фидбеку.

Neonik228 commented:
На текущий момент выполнил первые 2 задачи. На ум пришло 2 решения по поводу архитектуры базы данных (вот итоговая модель https://ibb.co/DGvQ8jK). У нас есть основная информация по ссылке и информация по изменениям (у каждого домена свои изменения и поля)
1) Сделать наследования таблиц от основной. В основной бы хранилась главная информация по типу имени, даты обновления. А в доп таблицы создать для каждого домена и там указать поля, которые хотим отслеживать (коммиты, комментарии и тд)  ведь они у всех разные.
2) Использовать одну таблицу, а всю отслеживаемую информацию хранить в jsonb.

Я выбрал 2 вариант. Ведь в первом случае для каждого домена нужно создать свою таблицу, и всякий раз делать 2 запроса (к основной и к дополнительной). Еще мне показалось трудность в том, что имена таблиц мы будем хардкодить, а не получать из запроса, чтобы сделать второй запрос. То есть если изменится имя таблицы, то придется идти и искать в коде, где оно использовалось и там тоже менять все. Во втором случае такого не будет. 
Буду рад фидбеку.

Gram3r commented on bot/src/main/java/ru/tinkoff/edu/java/bot/controllers/GeneralExceptionHandler.java:
```diff
@@ -0,0 +1,42 @@
+
+@RestControllerAdvice
+public class GeneralExceptionHandler extends ResponseEntityExceptionHandler {
+    private static final String DESCRIPTION_400 = "Invalid request parameters";
```
У нас ошибка 400 может быть не только с невалидным параметрами. Потом придется менять скорее всего

Gram3r commented on bot/src/test/java/ru/tinkoff/edu/java/logic/commands/InputHandlerTest.java:
```diff
@@ -0,0 +1,32 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class InputHandlerTest {
```
Вроде вся та же логика, что и у CommandTest, можно его назвать BaseTest и этот класс тоже от него отналседовать например

Gram3r commented on bot/src/test/java/ru/tinkoff/edu/java/logic/commands/ListCommandIT.java:
```diff
@@ -0,0 +1,50 @@
+import static org.mockito.Mockito.mockStatic;
+
+@ExtendWith(MockitoExtension.class)
+public class ListCommandIT extends CommandTest {
```
Название странное

Neonik-dev replied:
разве интеграционные тесты не так называются? Для плагина failsafe такие названия нужны

Gram3r commented on link-parser/src/test/java/ru/tinkoff/edu/java/parsers/GitHubParserTest.java:
```diff
@@ -0,0 +1,87 @@
+
+    @ParameterizedTest()
+    @ValueSource(strings = {
+            "https://github.com/Gram3r/Tinkoff-tourism",
```
Ах ты жук, хахахаха. А что это мы по-моему гиту лазаем))))

Neonik-dev replied:
ахахах, я сначала подумал, что ты рабочий проект случайно сделал не приватным, а публичным, удивился)

Gram3r commented on link-parser/src/test/java/ru/tinkoff/edu/java/parsers/StackOverFlowTest.java:
```diff
@@ -0,0 +1,81 @@
+        StackOverflowResponse response = (StackOverflowResponse) new App().main(link);
+
+        // then
+        assertAll(
```
А зачем в таких случаях ты юзаешь assertAll?

Gram3r reviewed (approved):
Хорошие исправления

Gram3r commented:
По поводу 5 дз я тебе к сожалению сейчас не отвечу, потому что не знаю что там сделать  с бд надо.

Neonik-dev commented:
> По поводу 5 дз я тебе к сожалению сейчас не отвечу, потому что не знаю что там сделать с бд надо.

случайно этот вопрос отправил сюда, а не на 5-ое дз
